### PR TITLE
feat: add stochastic_scoring_function for stochastic environments

### PR DESCRIPTION
**Pull request by manon-but-yes**
This PR adds an additional scoring function to `mdp_utils.py`: the `stochastic_scoring_function`, that wraps the `scoring_function` and allows the user to define a `play_reset_fn` function to reset the environment.
Setting `play_reset_fn` to `partial(env.reset)` allows to make the environment fully stochastic.

Initial tests using this new scoring function did not show important increase in computation time.

### CODE REVIEW

File: qdax/core/neuroevolution/mdp_utils.py [hunk 1]
```diff
@@ -179,6 +175,48 @@ def scoring_function(
...
+    play_reset_fn: Callable[[RNGKey], brax.envs.State],
+    play_step_fn: Callable[
+        [brax.envs.State, Params, RNGKey, brax.envs.Env],
+        Tuple[brax.envs.State, Params, RNGKey, QDTransition],
+    ],
+    behavior_descriptor_extractor: Callable[[QDTransition, jnp.ndarray], Descriptor],
+) -> Tuple[Fitness, Descriptor, Dict[str, Union[jnp.ndarray, QDTransition]], RNGKey]:
+    """Evaluates policies contained in policies_params in parallel.
+    The play_reset_fn function allows to define purely stochastic environments.
+    To use the reset function from the environment set it to "partial(env.reset)".
```
Comment (felixchalumeau) [inline qdax/core/neuroevolution/mdp_utils.py:200]: Why `partial(env.reset)`? Wouldn't `env.reset` work?

Comment (manon-but-yes) [reply qdax/core/neuroevolution/mdp_utils.py:200]: Yes `env.reset` would work indeed.
Also, is it useful to add to this comment that the user can also use it for the deterministic case by setting `play_reset_fn` such that it always returns the same `init_state`?

Comment (manon-but-yes) [reply qdax/core/neuroevolution/mdp_utils.py]: I changed the function name to "reset_based_scoring_function" following a discussion with Bryan, and updated its description.

**Comment by manon-but-yes** [comment]
To support the claim about computation time, here is the comparison of simple MAP-Elites on the Ant Omni against time in terms of QD-Score and Coverage. There are 3 baselines: the original `scoring_function`, the new `reset_based_scoring_function` using a deterministic `play_reset_fn` and the new `reset_based_scoring_function` using stochastic `play_reset_fn`.

I ran 3 replications on GPU. As mentioned in Jax documentation, on GPU even "deterministic" functions show some stochasticity, which explains the slight difference between replications of the two deterministic runs. 
The loss in performance in the stochastic environment is expected as random initialisation makes the task harder (offspring might behave totally different than their parents due to different initial conditions).

The time difference between the original and new stochastic scoring functions is approximately 5 seconds out of 140 seconds of run. In addition, this 5s difference propagates directly from a timing difference in the initialisation, probably due to some additional time spent jit-ing the new reset function.

Screenshot 2022-06-28 at 17 19 51

**Review by limbryan — APPROVED** [review]
